File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.cpp
class Solution {
public:
    int solve(int idx,int pre_idx, vector<int>& nums, int n, vector<vector<int>>& dp){
        if(idx == n) return 0;
        if(dp[idx][pre_idx+1] != -1) return dp[idx][pre_idx+1];

        int len = solve(idx+1, pre_idx, nums,n,dp);
        if(pre_idx == -1 || nums[pre_idx] < nums[idx]){ 
            len = max(len, 1 + solve(idx+1,idx,nums,n,dp));
        }

        return dp[idx][pre_idx+1] = len;
    }

    int lengthOfLIS(vector<int>& nums) {
        int n  = nums.size();
        vector<vector<int>> dp(n,vector<int>(n+1,-1));
        return solve(0,-1,nums,n,dp);
    }
};

File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var lengthOfLIS = function(nums) {
    let ans = 1
    const arr = []
    arr.push(nums[0])

    function lowerBound(nums, target){
        let l = 0, r = nums.length
        while(l<r){
            let m = l +Math.floor((r-l)/2)
            if(nums[m] < target){
                l = m+1;
            }
            else{
                r = m;
            }
        }
        return l;
    }
    for(let i=1;i<nums.length;i++){
        if(nums[i] > arr[arr.length-1]){
            ans++
            arr.push(nums[i])
        }
        else{
            const idx = lowerBound(arr, nums[i])
            arr[idx] = nums[i]
        }
    }
    return ans
};